src/net/connection.rs

```diff
@@ -1,17 +1,129 @@
-mod connection_pool;
 mod quality;
-mod timeout_thread;
 mod virtual_connection;
 
-pub use self::connection_pool::ConnectionPool;
 pub use self::quality::{NetworkQuality, RttMeasurer};
-pub use self::timeout_thread::TimeoutThread;
 pub use self::virtual_connection::VirtualConnection;
 
-use std::collections::HashMap;
-use std::net::SocketAddr;
-use std::sync::{Arc, RwLock};
+use crate::config::Config;
+use std::{collections::HashMap, net::SocketAddr, time::Duration};
 
-pub type Connection = Arc<RwLock<VirtualConnection>>;
-pub type Connections = HashMap<SocketAddr, Connection>;
-pub type ConnectionsCollection = Arc<RwLock<Connections>>;
+/// Maintains a registry of active "connections". Essentially, when we receive a packet on the
+/// socket from a particular `SocketAddr`, we will track information about it here.
+pub struct ActiveConnections {
+    connections: HashMap<SocketAddr, VirtualConnection>,
+}
+
+impl ActiveConnections {
+    pub fn new() -> Self {
+        Self {
+            connections: HashMap::new(),
+        }
+    }
+
+    /// Try to get a VirtualConnection by address. If the connection does not exist, it will be
+    /// inserted and returned.
+    pub fn get_or_insert_connection(
+        &mut self,
+        address: SocketAddr,
+        config: &Config,
+    ) -> &mut VirtualConnection {
+        self.connections
+            .entry(address)
+            .or_insert_with(|| VirtualConnection::new(address, config))
+    }
+
+    /// Removes the connection from `ActiveConnections` by socket address.
+    pub fn remove_connection(
+        &mut self,
+        address: &SocketAddr,
+    ) -> Option<(SocketAddr, VirtualConnection)> {
+        self.connections.remove_entry(address)
+    }
+
+    /// Check for and return `VirtualConnection`s which have been idling longer than `max_idle_time`.
+    pub fn idle_connections(&mut self, max_idle_time: Duration) -> Vec<SocketAddr> {
+        self.connections
+            .iter()
+            .filter(|(_, connection)| connection.last_heard() >= max_idle_time)
+            .map(|(address, _)| *address)
+            .collect()
+    }
+
+    /// Get the number of connected clients.
+    #[allow(dead_code)]
```

Comment:
Do we still need this allow?

Reply:
I think this was added early on for something but we haven't used it so I think it is fair to remove this function or at least this warning.

Reply:
Method is used in test code.